### PR TITLE
fix soundness bug with unsupported constraints

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102897

We do not raise constraint violations for complex binary conditions, such as conditions involving `%`. Moreover, while these constraints are discovered by our solver, the solver does not inject new constraint violations. This can result in cases where export passes, appropriate assertions are not added, and we get runtime crashes. 

Now, when the solver discovers constraints that are too complex, we force-specialize the involved dimensions and raise a constraint violation when such dimensions are marked dynamic. This forces the user to remove the dynamic marking, and causes the appropriate specialization assertions to be added.

Differential Revision: [D46415786](https://our.internmc.facebook.com/intern/diff/D46415786/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1686,7 +1689,35 @@ def raise_inconsistencies(self):
...
+            for s, exprs in self._univariate_inequalities.items()
+            if s not in self._substitutions
+        }
+        self._congruences = {
+            s: congruences
+            for s, congruences in self._congruences.items()
+            if s not in self._substitutions
+        }
+
+    def solve(self, disable_congruences=True):
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1720]: If and when we support specifying `%` constraints, we can enable this flag.

**Comment by avikchaudhuri** [comment]
Failing tests are caused by some mismatch in error format in https://github.com/pytorch/pytorch/pull/102666, not sure why these are not picked up by CI.

**Review by tugsbayasgalan — APPROVED** [review]

**Review by tugsbayasgalan — APPROVED** [review]

File: test/export/test_export.py [hunk 2]
```diff
@@ -213,6 +213,30 @@ def f(x, y):
...
+            torch._dynamo.exc.UserError,
+            (
+                "Some dynamic dimensions need to be specialized because "
+                "the constraints inferred for them are too complex to specify"
+                ".*\n.*\\[0\\], which was marked dynamic, must be specialized to 3"
+                ".*\n.*\\[1\\], which was marked dynamic, must be specialized to 4"
+            ),
+        ):
+            torch._export.export(m, (a,), constraints=constraints)
+        em = torch._export.export(m, (a,)).add_runtime_assertions()
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply test/export/test_export.py:234]: I recently landed a change that makes this private method.

[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply test/export/test_export.py:234]: Yeah, I'll wait for https://www.internalfb.com/diff/D46471787 to land internally before rebasing.

**Pull request closed without merging**